cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,192 @@
 
 #pragma once
 
+#include <cstdint>
 #include <cudf/strings/string_view.cuh>
+#include <hash/hash_constants.hpp>
+
+#include "cuda_runtime_api.h"
+#include "cudf/types.hpp"
+#include "driver_types.h"
+#include "vector_types.h"
 
 using hash_value_type = uint32_t;
 
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Core MD5 algorith implementation. Processes a single 512-bit chunk,
+ * updating the hash value so far. Does not zero out the buffer contents.
+ */
+void CUDA_DEVICE_CALLABLE md5_hash_step(md5_intermediate_data* hash_state,
+                                        md5_hash_constants_type const* hash_constants,
+                                        md5_shift_constants_type const* shift_constants)
+{
+  uint32_t A = hash_state->hash_value[0];
+  uint32_t B = hash_state->hash_value[1];
+  uint32_t C = hash_state->hash_value[2];
+  uint32_t D = hash_state->hash_value[3];
+
+  uint32_t* buffer_ints = reinterpret_cast<uint32_t*>(hash_state->buffer);
```

Comment:
This cast violates C++'s strict aliasing rules and is undefined behavior.

https://stackoverflow.com/questions/98650/what-is-the-strict-aliasing-rule

Reply:
Instead of using a cast, you should use `memcpy` to copy out 4B of the `hash_state->buffer` as a `uint32_t`. E.g.,

```
for (unsigned int j = 0; j < 64; j++) {
   uint32_t buffer_element_as_int;
   memcpy( &buffer_element_as_int, hash_state->buffer + j*sizeof(uint32_t),);
```